design/boto-design.md

```diff
@@ -0,0 +1,101 @@
+# Co-Existing with Boto design
+
+## Problem Description
+
+Existing code bases where paths are scattered throughout the codebase (which is rather common for DE/DS code) have a hard time incorporating the lakeFS S3 gateway for several reasons:
+1. Clients (namely, boto3 and pandas) don't support multiple S3 endpoints with differing credentials. This makes it hard to change only a subset of the data to be in lakeFS because the client, once initialized, can talk to either lakeFS or S3 but not both.
+2. Once data is migrated from S3 to lakeFS, even if the repo has the same name as the previous bucket, it's still not a drop-in replacement, because in lakeFS you'd also need to specify the version ref (branch/commit/tag) you're using, as a prefix.
+
+Solving 1 is hard because it requires a big infrastructure change: instead of using a single initialized client (sometimes it's a default or global one), now multiple clients need to be introduced (one for S3, one for lakeFS) with custom logic for when to use which one.
+Solving 2 is hard because it involves changing many paths scattered throughout the code base: some are dynamically generated by the code, some come from env vars or configuration files - it's not always easy to ensure coverage.
+
+## Goals
+* Make it easy for new users to transfer existing data to lakefs
+* Isolation - Users should be able to separate access to bucket/prefix between lakefs and s3
+* Users should do the minimal set of changes possible in code to set lakefs
+
+## Non Goals
+* Provide new features that weren't previously possible through lakefs
+
+## Value Propositions
+When a user with an existing python code that uses boto3 client to access s3 wants to move only part of its buckets/prefixes to lakefs, botwo allows using lakefs and s3 together with minimum user's code changes.
+
+## Proposed Design
+The botwo package provides a boto3 like client with two boto3 clients.
+It handles several things for the user:
+* Holds two boto clients and routes the requests between them by the bucket/prefix configuration.
+* Maps 1:1 with the boto AWS service API. All botwo client AWS functions will be generated on the fly according to boto3 client AWS functions.
+* Holds special treatment for two buckets functions like `upload_part_copy`. If more two buckets functions will be added to s3 in the future, we will need to implement them as well.
+* It has more functionality like the event-system that will be implemented by using one of its clients (the default one).
+
+## Alternatives Considered
+
+### Route the requests to lakefs endpoint by the prefix
+```
+s3 = boto3.client('s3')
+if bucket == "bucket_a" and prefix == "my_path":
+  client.endpoint_url  = 'https://lakefs.example.com'
+```
+Does not work because once you initialize boto client with parameters you can’t change them in run time.
+
+### Method Overriding
+
+Monkey patching / overriding base class functions in boto3 with inheritance.
```

Comment:
What about wrapping using some explicit AOP (aspect-oriented programming) method?

Reply:
I will check it out.